aws/s3.go

```diff
@@ -0,0 +1,295 @@
+package aws
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/gruntwork-io/cloud-nuke/logging"
+	"github.com/gruntwork-io/gruntwork-cli/errors"
+)
+
+// getS3BucketRegion returns S3 Bucket region.
+func getS3BucketRegion(svc *s3.S3, bucketName *string) (string, error) {
+	input := &s3.GetBucketLocationInput{
+		Bucket: bucketName,
+	}
+
+	result, err := svc.GetBucketLocation(input)
+	if err != nil {
+		return "", err
+	}
+
+	if result.LocationConstraint == nil {
+		// GetBucketLocation returns nil for us-east-1
+		// https://github.com/aws/aws-sdk-go/issues/1687
+		return "us-east-1", nil
+	}
+	return *result.LocationConstraint, nil
+}
+
+// getS3BucketTags returns S3 Bucket tags.
+func getS3BucketTags(svc *s3.S3, bucketName *string) ([]map[string]string, error) {
+	input := &s3.GetBucketTaggingInput{
+		Bucket: bucketName,
+	}
+
+	tags := []map[string]string{}
+
+	result, err := svc.GetBucketTagging(input)
+	if err != nil {
+		return tags, err
+	}
+
+	for _, tagSet := range result.TagSet {
+		tags = append(tags, map[string]string{"Key": *tagSet.Key, "Value": *tagSet.Value})
+	}
+
+	return tags, err
+}
+
+// getAllS3Buckets lists and returns a map of per region AWS S3 buckets which were created before excludeAfter
+func getAllS3Buckets(session *session.Session, excludeAfter time.Time, bucketNameSubStr string) (map[string][]*string, error) {
+	svc := s3.New(session)
+
+	input := &s3.ListBucketsInput{}
+
+	output, err := svc.ListBuckets(input)
+	if err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
+
+	var bucketNamesPerRegion = make(map[string][]*string)
+
+OUTER:
+	for _, bucket := range output.Buckets {
+		logging.Logger.Debugf("Checking - Bucket %s", *bucket.Name)
+
+		if len(bucketNameSubStr) > 0 {
+			if !strings.Contains(*bucket.Name, bucketNameSubStr) {
+				logging.Logger.Debugf("Skipping - Bucket %s - failed substring filter - %s", *bucket.Name, bucketNameSubStr)
+				continue
+			}
+		}
+
+		if !excludeAfter.After(*bucket.CreationDate) {
+			logging.Logger.Debugf("Skipping - Bucket %s - matched CreationDate filter", *bucket.Name)
+			continue
+		}
+
+		bucketTags, err := getS3BucketTags(svc, bucket.Name)
+		if len(bucketTags) > 0 {
+			for _, tagSet := range bucketTags {
+				if tagSet["Key"] == "cloud-nuke-excluded" && tagSet["Value"] == "true" {
```

Comment:
Can you add a comment here indicating why we are doing this, and include a link to my comment? (https://github.com/gruntwork-io/cloud-nuke/pull/96#pullrequestreview-367772274)

Also if you have time, adding a section in the README documenting this for discoverability would be great!

Reply:
Will add comment. This is already documented in the README as ```Deleting all S3 buckets in an AWS account - except for buckets tagged with Key=cloud-nuke-excluded Value=true
```